### PR TITLE
Allow to define an object storage endpoint for package registry [feature]

### PR DESCRIPTION
**Pull request by Akanoa**
Hello :)

I'm using since more than one year. Really useful.

But it misses something, although the gitlab-ce project allows to configure the object storage for package registry. The current installation doesn't give this possibility.

My pull-request has as aim to allow user to define all the settings related to the `packages` section of the `gitlab.yml` file.

Beside of this I also factorized some settings that can be shared across all storage section.

I introduce then some more variable, with taking care to do not break previous behaviors regarding the default values:

- `AWS_REGION`  AWS Region. Defaults to `us-east-1`
- `AWS_HOST`  Configure this for an compatible AWS host like minio. Defaults to `$AWS_HOST`. Defaults to `s3.amazon.com`
- `AWS_ENDPOINT` AWS Endpoint like `http://127.0.0.1:9000`. Defaults to `nil`
- `AWS_PATH_STYLE`  Changes AWS Path Style to 'host/bucket_name/object' instead of 'bucket_name.host/object'. Defaults to `true` 

Here the list of all new variables:

- `GITLAB_PACKAGES_ENABLED` Enable/Disable Pakages support. Defaults to `true`.
- `GITLAB_PACKAGES_DIR` Directory to store the packages data. Defaults to `$GITLAB_SHARED_DIR/packages`
- `GITLAB_PACKAGES_OBJECT_STORE_ENABLED` Enables Object Store for Packages that will be remote stored. Defaults to `false`
- `GITLAB_PACKAGES_OBJECT_STORE_REMOTE_DIRECTORY` Bucket name to store the packages. Defaults to `packages`
- `GITLAB_PACKAGES_OBJECT_STORE_DIRECT_UPLOAD` Set to true to enable direct upload of Packages without the need of local shared storage.  Defaults to `false`
- `GITLAB_PACKAGES_OBJECT_STORE_BACKGROUND_UPLOAD` Temporary option to limit automatic upload. Defaults to `false`
- `GITLAB_PACKAGES_OBJECT_STORE_PROXY_DOWNLOAD` Passthrough all downloads via GitLab instead of using Redirects to Object Storage. Defaults to `false` 
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_PROVIDER` Connection Provider for the Object Store. (`AWS` or `Google`) Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_PROVIDER` (`AWS`)
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_ACCESS_KEY_ID` AWS Access Key ID for the Bucket. Defaults to `$AWS_ACCESS_KEY_ID`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_SECRET_ACCESS_KEY` AWS Secret Access Key. Defaults to `$AWS_SECRET_ACCESS_KEY`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_REGION` AWS Region. Defaults to `$AWS_REGION`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_HOST` Configure this for an compatible AWS host like minio. Defaults to `$AWS_HOST`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_ENDPOINT` AWS Endpoint like `http://127.0.0.1:9000`. Defaults to `$AWS_ENDPOINT`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_PATH_STYLE`  Changes AWS Path Style to 'host/bucket_name/object' instead of 'bucket_name.host/object'. Defaults to `AWS_PATH_STYLE`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_GOOGLE_PROJECT` Google project. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_PROJECT`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_GOOGLE_CLIENT_EMAIL` Google service account. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_CLIENT_EMAIL`
- `GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_GOOGLE_JSON_KEY_LOCATION` Default Google key file. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_JSON_KEY_LOCATION` (`/gcs/key.json`)

Thanks

### CODE REVIEW

**Comment by sameersbn** [comment]
hello @sachilles, the changes in this PR look good to me. could you please also take a look?

**Review by sameersbn — APPROVED** [review]

**Comment by sachilles** [comment]
@sameersbn I will do that. But I will not be able to do it before tomorrow.

File: assets/runtime/config/gitlabhq/gitlab.yml [hunk 1]
```diff
@@ -320,25 +320,33 @@ production: &base
...
-        # aws_signature_version: 4 # For creation of signed URLs. Set to 2 if provider does not support v4.
-        # path_style: true # Use 'host/bucket_name/object' instead of 'bucket_name.host/object'
-
+        provider: {{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_PROVIDER}} # Only AWS supported at the moment
+        #start-packages-aws
+        aws_access_key_id: {{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_ACCESS_KEY_ID}}
+        aws_secret_access_key: {{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_SECRET_ACCESS_KEY}}
+        region: {{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_REGION}}
+        host: '{{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_HOST}}' # default: s3.amazonaws.com
+        aws_signature_version: 4 # For creation of signed URLs. Set to 2 if provider does not support v4.
```
Comment (sachilles) [inline assets/runtime/config/gitlabhq/gitlab.yml:339]: Wouldn't it be reasonable if the value of `aws_signature_version` would be set via variables from the file `env-defaults`? Possibly also via environment variable from user given by docker or docker-compose?

Comment (sachilles) [reply assets/runtime/config/gitlabhq/gitlab.yml:339]: Should the other settings related to S3-compatible connection setting (see https://docs.gitlab.com/ee/administration/object_storage.html#s3-compatible-connection-settings) be ignored?

File: assets/runtime/config/gitlabhq/gitlab.yml [hunk 2]
```diff
@@ -320,25 +320,33 @@ production: &base
...
+        endpoint: '{{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_ENDPOINT}}' # default: nil - Useful for S3 compliant services such as DigitalOcean Spaces
+        path_style: {{GITLAB_PACKAGES_OBJECT_STORE_CONNECTION_AWS_PATH_STYLE}} # Use 'host/bucket_name/object' instead of 'bucket_name.host/object'
+        #end-packages-aws
+        #start-packages-gcs
```
Comment (sachilles) [inline assets/runtime/config/gitlabhq/gitlab.yml:343]: I haven't used Google Cloud Storage yet, but following https://docs.gitlab.com/ee/administration/object_storage.html#google-cloud-storage-gcs there is another variable (`google_application_default`). Is this not required?

**Review by sachilles — COMMENTED** [review]
Hi, so far I had nothing to do with object storage.  Therefore, I can basically only comment on gitlab's documentation and how the given implementation is done. So far, I hope that my comments help.

In principle, the commit looks good. (But I guess, before merging there should be a rebase, isn't it?)

**Comment by sameersbn** [comment]
@Akanoa could you please rebase? and comment on the review

**Comment by Akanoa** [comment]
Yes I'll do it, and I'll make asked modifications, I am agree with them. 

Thanks for the code review. :)

**Comment by sachilles** [comment]
@Akanoa Could be so kind to reset the branch `feature/allow-to-define-packages-object-storage` to commit `012c2ff` and use `git rebase` instead of merging the master to your feature branch.

**Comment by Akanoa** [comment]
Oh sorry I've used the github "fetch upstream" button.

**Comment by Akanoa** [comment]
Hello @sachilles,

I'll do another pull-request to add the support about additional settings regarding Google Cloud Storage.

Regards,

**Comment by Akanoa** [comment]
Hello @sameersbn,

Thank you very much for merging my participation.

**Pull request merged**